File: output_Codeium/cpp/Medium/623-generate-parentheses.cpp
/**https://leetcode.com/problems/generate-parentheses/ */
//Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
// 
//Example 1:
//Input: n = 3
//Output: ["((()))","(()())","(())()","()(())","()()()"]
//Example 2:
//Input: n = 1
//Output: ["()"]
// 
//Constraints:
//	1 <= n <= 8
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        backtrack(res, "", 0, 0, n);
        return res;
    }
    
    void backtrack(vector<string>& res, string cur, int open, int close, int n) {
        if (cur.size() == n * 2) {
            res.push_back(cur);
            return;
        }
        
        if (open < n)
            backtrack(res, cur + "(", open + 1, close, n);
        if (close < open)
            backtrack(res, cur + ")", open, close + 1, n);
    }
};

File: output_Codeium/python/Medium/623-generate-parentheses.py
#https://leetcode.com/problems/generate-parentheses/
#Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
# 
#Example 1:
#Input: n = 3
#Output: ["((()))","(()())","(())()","()(())","()()()"]
#Example 2:
#Input: n = 1
#Output: ["()"]
# 
#Constraints:
#	1 <= n <= 8
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        result = []
        self.generate(result, "", 0, 0, n)
        return result
    
    def generate(self, result, current, openCount, closeCount, n):
        if len(current) == n * 2:
            result.append(current)
            return
        
        if openCount < n:
            self.generate(result, current + "(", openCount + 1, closeCount, n)
        if closeCount < openCount:
            self.generate(result, current + ")", openCount, closeCount + 1, n)
